clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,157 @@
+use crate::utils::{is_direct_expn_of, span_lint};
+use core::ops::Deref;
+use if_chain::if_chain;
+use matches::matches;
+use rustc::hir::{Expr, ExprKind, Guard, Mutability, StmtKind, UnOp};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint, ty};
+use syntax_pos::Span;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function/method calls with a mutable
+    /// parameter in `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!` macros.
+    ///
+    /// **Why is this bad?** In release builds `debug_assert!` macros are optimized out by the
+    /// compiler.
+    /// Therefore mutating something in a `debug_assert!` macro results in different behaviour
+    /// between a release and debug build.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// debug_assert_eq!(vec![3].pop(), Some(3));
+    /// // or
+    /// fn take_a_mut_parameter(_: &mut u32) -> bool { unimplemented!() }
+    /// debug_assert!(take_a_mut_parameter(&mut 5));
+    /// ```
+    pub DEBUG_ASSERT_WITH_MUT_CALL,
+    correctness,
+    "mutable arguments in `debug_assert{,_ne,_eq}!`"
+}
+
+declare_lint_pass!(DebugAssertWithMutCall => [DEBUG_ASSERT_WITH_MUT_CALL]);
+
+fn check_for_mutable_call(cx: &LateContext<'_, '_>, span: Span, e: &Expr) -> Option<Span> {
```

Comment:
is there a reason why this can't be written as a visitor?

Reply:
Already done. Didn't know about the Visitor :zipper_mouth_face:

Reply:
It is now. I didn't realized, that @hellow554 tried to implement a visitor, but thought this was part of the `debug_assert` match or something, until c1d2ed1. I could've saved him a lot of work... :smile: